Miscellaneous improvements and fixes concerning the endpoints UID checks and surrounding logic

giorio94 opened the pull request:
Please review commit by commit, and refer to the individual descriptions for additional details. 

I'm currently a bit on the fence whether any commit would need to be backported to earlier branches, although I'm leaning more towards not backporting them given that they address very specific corner case situations. The only issue which looks a bit more problematic to me is the one fixed in https://github.com/cilium/cilium/pull/36392/commits/6bebde0ce4700556edef93bb1a6f56491fa5986b, because it could lead to accessing a nil variable, but it is my understanding that the UID is always guaranteed to be set. Anyways, please let me know your thoughts during the review.

Related: https://github.com/cilium/cilium/pull/36292#discussion_r1865704978 

/cc @squeed

giorio94 committed:
endpoint: simplify handling of no-track-rules/bw-policy

Currently, the update of no track rules and bandwidth policies
is handled posting an event to the event queue, which in turn
executes a callback to retrieve the corresponding pod from the
store, and in turn the annotation itself. Let's simplify this
by directly propagating the annotation(s) in the event itself,
given that all callers have access to pod object, and subsequent
annotation updates would just trigger a new event. This also
prevents a possible incorrect ending state in case the same
annotation is flipped multiple times.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
endpoint: simplify handling of mismatching pod UID

Move the UID mismatch check directly to the function that retrieves the
metadata for a given endpoint, and let it return an error in that case.
This change ensures that the UID is always properly checked, and allows
simplifying the error handling processing in the callers.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
k8s: simplify signature of GetPodMetadata

The function always returns a nil error, hence let's just drop it.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
endpoint: add unit test covering handleOutdatedPodInformer

Introduce a new unit test to cover the handleOutdatedPodInformer
function, in preparation for the subsequent changes. In particular,
this proves two problems currently affecting it:

1. The function does not propagate the error returned by the fetch
   function if the endpoint UID is not set.
2. The function does not perform any retry in case of pod UID
   mismatch, conversely to the intended behavior.

--- FAIL: TestHandleOutdatedPodInformer (0.00s)
    --- FAIL: TestHandleOutdatedPodInformer/pod_not_found_(epUID:_) (0.00s)
        endpoint_test.go:278:
                Error Trace:    .../daemon/cmd/endpoint_test.go:278
                Error:          Not equal:
                                expected: *errors.StatusError(&errors.StatusError{...
                                actual  : <nil>(<nil>)
                Test:           TestHandleOutdatedPodInformer/pod_not_found_(epUID:_)
    --- FAIL: TestHandleOutdatedPodInformer/uid_mismatch_(epUID:_uid) (0.00s)
        endpoint_test.go:288:
                Error Trace:    .../daemon/cmd/endpoint_test.go:288
                Error:          Not equal:
                                expected: 0x14
                                actual  : 0x1
                Test:           TestHandleOutdatedPodInformer/uid_mismatch_(epUID:_uid)
                Messages:       Incorrect number of retries
    --- FAIL: TestHandleOutdatedPodInformer/uid_mismatch,_then_resolved_(epUID:_uid) (0.00s)
        endpoint_test.go:278:
                Error Trace:    .../daemon/cmd/endpoint_test.go:278
                Error:          Not equal:
                                expected: <nil>(<nil>)
                                actual  : *errors.errorString(&errors.errorString{s:"pod store outdated"})
                Test:           TestHandleOutdatedPodInformer/uid_mismatch,_then_resolved_(epUID:_uid)
        endpoint_test.go:288:
                Error Trace:    .../daemon/cmd/endpoint_test.go:288
                Error:          Not equal:
                                expected: 0x6
                                actual  : 0x1
                Test:           TestHandleOutdatedPodInformer/uid_mismatch,_then_resolved_(epUID:_uid)
                Messages:       Incorrect number of retries
FAIL
FAIL    github.com/cilium/cilium/daemon/cmd     1.336s
FAIL

Related: f6606c6ccdb8 ("daemon,endpoint,cni: Pass pod UID through CNI ADD")
Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
endpoint: let handleOutdatedPodInformer propagate fetch errors

The handleOutdatedPodInformer function currently does not propagate
fetch errors if the endpoint UID is not set. Although errors can only
occur if either the namespace or the pod are not present yet in the
store, let's get this fixed, to prevent possibly accessing uninitialized
variables.

Fixes: f6606c6ccdb8 ("daemon,endpoint,cni: Pass pod UID through CNI ADD")
Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
endpoint: let handleOutdatedPodInformer correctly retry on uid mismatch

According to the description of the blamed commit:

    If the UID is outdated, this means that Cilium has not yet received the
    updated pod event that corresponds with the CNI ADD. In this case,
    Cilium will attempt to query the pod store up for to 2 seconds. If the
    pod store does not have the altest pod reference after 2 seconds, Cilium
    will trigger a direct fetch from the apiserver [...]

However, the current implementation never performs any retry, because
the underlying retry logic returns immediately in case an error is
returned by the function. Let's fix this by not returning an error in
that case, as well as changing the duration period to actually match
the expected max retry time of 2 seconds.

Fixes: f6606c6ccdb8 ("daemon,endpoint,cni: Pass pod UID through CNI ADD")
Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
endpoint: compute updated metadata on pod UID mismatch

The blamed commit introduced a fallback logic to retrieve the pod
corresponding to a given endpoint directly from k8s if the store
is out of date, and contains an entry with a mismatching UID. However,
the metadata is not recomputed based on the newly retrieved pod,
hence potentially configuring incorrect values. Let's get this fixed.

Fixes: f6606c6ccdb8 ("daemon,endpoint,cni: Pass pod UID through CNI ADD")
Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

squeed commented on daemon/cmd/endpoint.go:
```diff
@@ -178,15 +178,25 @@ func getEndpointIDHandler(d *Daemon, params GetEndpointIDParams) middleware.Resp
+	return p, metadata, err
+}
+
+func (d *Daemon) fetchK8sMetadataForEndpointFromPod(p *slim_corev1.Pod) (*endpoint.K8sMetadata, error) {
```
stylistic nit: what if this function took the full namespace? Then it no longer needs to return `error`.

squeed replied:
Aha, I see it's referenced below. Feel free to disregard.

giorio94 replied:
Yep, the reason is that it is now also directly called by `createEndpoint` after retrieving the pod from k8s, where we don't have the namespace object already available.

giorio94 commented on daemon/cmd/endpoint.go:
```diff
@@ -177,13 +177,18 @@ func getEndpointIDHandler(d *Daemon, params GetEndpointIDParams) middleware.Resp
 		return nil, nil, err
 	}
 
+	if uid != "" && uid != string(p.GetUID()) {
+		return nil, nil, podStoreOutdatedErr
+	}
```
A stricter approach would be moving this check to `fetchK8sMetadataForEndpointFromPod` below, so that it gets performed also after directly retrieving the pod object from k8s in the fallback path of `createEndpoint`.

giorio94 committed:
endpoint: drop redundant annotations from endpoint metadata

All callers either ignore this field or have direct access to the full
pod object. Hence, let's simply remove it, to prevent possible ambiguity.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

christarazi reviewed (approved):
Nice, thanks for simplifying and fixing this code
